nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -256,6 +375,312 @@ impl DataStore {
+    }
+
+    // TODO: separate this operation from update so that we can have /link 409
+    // or whatever when the association already exists?
+    pub async fn ip_pool_associate_resource(
+        &self,
+        opctx: &OpContext,
+        ip_pool_resource: IpPoolResource,
+    ) -> CreateResult<IpPoolResource> {
+        use db::schema::ip_pool_resource::dsl;
+        opctx
+            .authorize(authz::Action::CreateChild, &authz::IP_POOL_LIST)
+            .await?;
+
+        diesel::insert_into(dsl::ip_pool_resource)
+            .values(ip_pool_resource.clone())
+            // We have two constraints that are relevant here, and we need to
+            // make this behave correctly with respect to both. If the entry
+            // matches an existing (ip pool, silo/fleet), we want to update
+            // is_default because we want to handle the case where someone is
+            // trying to change is_default on an existing association. But
+            // you can only have one default pool for a given resource, so
+            // if that update violates the unique index ensuring one default,
+            // the insert should still fail.
+            // note that this on_conflict has to have all three because that's
+            // how the pk is defined. if it only has the IDs and not the type,
+            // CRDB complains that the tuple doesn't match any constraints it's
+            // aware of
+            .on_conflict((
+                dsl::ip_pool_id,
+                dsl::resource_type,
+                dsl::resource_id,
+            ))
+            .do_update()
+            .set(dsl::is_default.eq(ip_pool_resource.is_default))
+            .returning(IpPoolResource::as_returning())
+            .get_result_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel(
+                    e,
+                    ErrorHandler::Conflict(
+                        ResourceType::IpPoolResource,
+                        &format!(
+                            "ip_pool_id: {:?}, resource_id: {:?}, resource_type: {:?}",
+                            ip_pool_resource.ip_pool_id,
+                            ip_pool_resource.resource_id,
+                            ip_pool_resource.resource_type,
+                        )
+                    ),
+                )
+            })
+    }
+
+    // TODO: make default should fail when the association doesn't exist.
+    // should it also fail when it's already default? probably not?
+    pub async fn ip_pool_set_default(
+        &self,
+        opctx: &OpContext,
+        authz_ip_pool: &authz::IpPool,
+        authz_silo: &authz::Silo,
+        is_default: bool,
+    ) -> UpdateResult<IpPoolResource> {
+        use db::schema::ip_pool_resource::dsl;
+
+        // TODO: correct auth check
+        opctx
+            .authorize(authz::Action::CreateChild, &authz::IP_POOL_LIST)
+            .await?;
```

Comment:
Pretty much all the auth checks in this PR are weird. In practice I think this works simply because you need fleet admin to do any of this stuff, and this is effectively checking for fleet admin. But we may need another synthetic resource representing IP pool/silo links to make this a bit more clear.

Reply:
So you have a few options in these. The _correct_ way may likely be to create a dedicate authz resource representing pool/silo links and auth check that. Alternatively I think if you just checked modify on both resources that would be sufficient. I wouldn't do create child though.